infobind/cache.go

```diff
@@ -0,0 +1,224 @@
+	"fmt"
+	"sync/atomic"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/sqlexec"
+	log "github.com/sirupsen/logrus"
+)
+
+const defaultBindCacheSize = 5
+
+// BindData store the basic bind info and bindSql astNode.
+type BindData struct {
+	bindRecord
+	ast ast.StmtNode
+}
+
+// BindCache hold a bindDataMap, key:origin sql hash value: bindData slice.
+type BindCache struct {
+	Cache map[string][]*BindData
+}
+
+// Handle hold a atomic bindCache.
+type Handle struct {
+	bind atomic.Value
+}
+
+// HandleUpdater use to update the BindCache.
+type HandleUpdater struct {
+	Parser         *parser.Parser
+	LastUpdateTime types.Time
+	Ctx            sessionctx.Context
+	*Handle
+}
+
+type bindRecord struct {
+	OriginalSQL string
+	BindSQL     string
+	Db          string
+	Status      int64
+	CreateTime  types.Time
+	UpdateTime  types.Time
+	Charset     string
+	Collation   string
+}
+
+// NewHandle create a Handle with a BindCache.
+func NewHandle() *Handle {
+	handle := &Handle{}
+	return handle
+}
+
+// Get get bindCache from a Handle.
+func (h *Handle) Get() *BindCache {
+	bc := h.bind.Load()
+
+	if bc != nil {
+		return bc.(*BindCache)
+	}
+
+	return nil
+}
+
+// LoadDiff use to load new bind info to bindCache bc.
+func (h *HandleUpdater) loadDiff(sql string, bc *BindCache) error {
+	tmp, err := h.Ctx.(sqlexec.SQLExecutor).Execute(context.Background(), sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	rs := tmp[0]
+	defer terror.Call(rs.Close)
+
+	fs := rs.Fields()
+	chk := rs.NewChunk()
+	for {
+		err = rs.Next(context.TODO(), chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if chk.NumRows() == 0 {
+			return nil
+		}
+		it := chunk.NewIterator4Chunk(chk)
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			record, err := decodeBindTableRow(row, fs)
+			if err != nil {
+				log.Errorf("row decode error %s", err)
+				continue
+			}
+			err = bc.appendNode(h.Ctx, record, h.Parser)
+			if err != nil {
+				continue
+			}
+
+			if record.UpdateTime.Compare(h.LastUpdateTime) == 1 {
+				h.LastUpdateTime = record.UpdateTime
+			}
+		}
+		chk = chunk.Renew(chk, h.Ctx.GetSessionVars().MaxChunkSize)
+	}
+}
+
+// Update update the HandleUpdater's bindCache if tidb first startup,the fullLoad is true,otherwise fullLoad is false.
+func (h *HandleUpdater) Update(fullLoad bool) error {
+	var (
+		err error
+		sql string
+	)
+	bc := h.Get()
+
+	length := defaultBindCacheSize
+	if bc != nil {
+		length = len(bc.Cache)
+	}
+
+	newBc := &BindCache{
+		Cache: make(map[string][]*BindData, length),
+	}
+
+	if bc != nil {
+		for hash, bindDataArr := range bc.Cache {
+			newBc.Cache[hash] = append(newBc.Cache[hash], bindDataArr...)
+		}
+	}
+
+	if fullLoad {
+		sql = fmt.Sprintf("select * from mysql.bind_info")
```

Comment:
It is better to use sql like `select a, b, c from t` to avoid compatibility issues when we change the table schema in the future.

Reply:
ok